### PR TITLE
[TOPI][AutoTVM] NHWC conv2d templates for ARM

### PR DESCRIPTION
**Pull request by jackwish**
Per https://github.com/dmlc/tvm/pull/3754 and https://github.com/dmlc/tvm/pull/3141#issuecomment-526434398 , we are enabling NHWC conv2d templates for ARM as a *nearly* final solution. The benefits include:
* Enable NHWC schedule directly. Previously, we need to transpose between NCHW and NHWC.
* AutoTVM now can tune NHWC directly. Previously, we need to build a NCHW network to tune.
* Potential performance advantage in NHWC which known to community.

Cowork with @FrozenGene and @etaf .

This is a **draft** to loop people who may have interest, @anijain2305 @tmoreau89 . Will loop more when the PR is ready, thank you. :)

### CODE REVIEW

**Comment by anijain2305** [comment]
@jackwish Very glad to see this :)
Does this also support int8 computation?

**Comment by jackwish** [comment]
Hi Animesh, thanks for the interest, INT8 support will be added later, we'd
like to improve the draft schedule first :)

Animesh Jain <notifications@github.com> 于2019年8月30日周五 上午11:53写道：

**Comment by tmoreau89** [comment]
Great to see this draft @jackwish ! A question I have regarding code maintainability: should we consider reusing some common infrastructure between x86 and ARM?

**Comment by kevinthesun** [comment]
@jackwish Glad to see this! 

As we discussed in https://github.com/dmlc/tvm/pull/3754, NWHC layout can further boost conv2d on arm_cpu. In this case, we can use NHWC by default as backend schedule for arm_cpu. Combining with [enhanced alter op layout pass
](https://github.com/dmlc/tvm/issues/3670), we can boost models, such as MXNet model, which have NCHW as data layout.

File: topi/python/topi/arm_cpu/conv2d.py [hunk 1]
```diff
@@ -132,177 +143,52 @@ def _callback(op):
     traverse_inline(s, outs[0].op, _callback)
     return s
 
+@autotvm.register_topi_schedule(schedule_conv2d_nhwc, 'arm_cpu', ['direct'])
+def schedule_conv2d_nchw_arm_cpu(cfg, outs):
```
Comment (tmoreau89) [inline topi/python/topi/arm_cpu/conv2d.py]: rename to `schedule_conv2d_nhwc_arm_cpu`

**Comment by FrozenGene** [comment]
@tmoreau89 yes, I agree. we could consider opening a RFC to discuss. Many code / schedule can be unified between x86 with arm. However, There is some different places we will have. Especially tensorize for producing special instruction.

**Comment by jackwish** [comment]
> Great to see this draft @jackwish ! A question I have regarding code maintainability: should we consider reusing some common infrastructure between x86 and ARM?

Yes @tmoreau89 , that is really a great idea to share code between x86 and ARM (or any other CPU ISA). And, we have some internal discussions previously. Personally, I am thinking that have a module which handles generic CPU schedules, `topi.cpu` for example, and let the arch-dependent code lies in `topi.{isa}`. We can further discuss this, I was planning to draft a RPC next week, what you say :)

**Comment by jackwish** [comment]
> @jackwish Glad to see this!
> 
> As we discussed in #3754, NWHC layout can further boost conv2d on arm_cpu. In this case, we can use NHWC by default as backend schedule for arm_cpu. Combining with [enhanced alter op layout pass ](https://github.com/dmlc/tvm/issues/3670), we can boost models, such as MXNet model, which have NCHW as data layout.

Yes @kevinthesun , we are working on this to improve performance.

File: topi/python/topi/arm_cpu/conv2d.py (hunk 1 above)
Comment (jackwish) [reply topi/python/topi/arm_cpu/conv2d.py]: Yes, thank you!

**Comment by tmoreau89** [comment]
@jackwish an RFC sounds like a great idea. I agree, there will be differences between different backends (tensorization for ARM 8bit, and bitserial; vectorize for AVX etc.) but overall the schedules shouldn't diverge too much. It will make maintaining the code bases more tenable; and minimize technical debt in the long run

**Comment by jackwish** [comment]
Hi guys, I am a bit busy this week, will come back to this (for the RFC part) later :)

**Comment by yzhliu** [comment]
Is NHWC generally faster than NCHW on ARM CPUs? should we transform NCHW to NHWC in `alter_layout`?

**Comment by jackwish** [comment]
Hi Yizhi, thank you for ask. NHWC is not generally faster than NCHW, it
depends on the schedule still. I guess we can use AutoTVM to decide which
layout is preferred for specific workloads :)

Regards

Yizhi Liu <notifications@github.com> 于2019年9月9日周一 上午11:06写道：

**Comment by jwfromm** [comment]
@jackwish, are you sure NHWC isn't faster than NCHW on ARM? In my experience I've consistently observed speeds up to 5X faster and haven't found any cases where NCHW is preferable. I'm definitely on in support of automatically altering to NHWC for arm_cpu workloads.

**Comment by jackwish** [comment]
> @jackwish, are you sure NHWC isn't faster than NCHW on ARM? In my experience I've consistently observed speeds up to 5X faster and haven't found any cases where NCHW is preferable. I'm definitely in support of automatically altering to NHWC for arm_cpu workloads.

Hi @jwfromm , sorry if I made any confusing comments, but I was saying that NHWC is *not always* faster than NCHW. To achieve better performance, sound schedule is needed.

It's really impressive to obtain 5X performance for NHWC over NCHW, which I think this PR cannot make it happen (sad story)... Would you please share more your design if that's is possible? I am sure it must be inspiring and insightful!

**Comment by jwfromm** [comment]
I've been using the NHWC schedule written by @cowanmeg, which can be found in the quantized_end2end branch of her TVM fork (https://github.com/cowanmeg/tvm/tree/quantized_end2end). Maybe she can chime in on what works well for NHWC.

**Comment by jackwish** [comment]
> I've been using the NHWC schedule written by @cowanmeg, which can be found in the quantized_end2end branch of her TVM fork (https://github.com/cowanmeg/tvm/tree/quantized_end2end). Maybe she can chime in on what works well for NHWC.

Thank you @jwfromm. I have taken a look at [the code](https://github.com/cowanmeg/tvm/blob/quantized_end2end//topi/python/topi/arm_cpu/conv2d.py#L369), which is very similar to this PR. Still, I am very curious about the 5X performance, would you please share the detail of getting the performance (I will try @cowanmeg 's schedule too)? As mobilenet1 with NCHW can get 121 ms on Raspberry Pi 3B, 5X means ~25 ms, which outperforms many *fast* quantization optimization approaches I think.

**Comment by jwfromm** [comment]
Hi Jack,

I haven't tried benchmarking mobilenet. Because it uses separable convolutions instead of the more typical convolution layers in the models I've worked with it's very reasonable that data layout format might have a much smaller impact. Most recently I found that SqueezeNet in NCHW format takes 780ms to run in NCHW format and 182ms to run in NHWC format. It might be worth checking out your schedules on a squeezenet or vggnet (lots of standard convolutions).

**Comment by ZihengJiang** [comment]
Hey @jackwish , nice work for supporting NHWC data layout. It would be great if you could give some performance number of common models on ARM CPUs.

**Comment by jackwish** [comment]
> Hi Jack,
> 
> I haven't tried benchmarking mobilenet. Because it uses separable convolutions instead of the more typical convolution layers in the models I've worked with it's very reasonable that data layout format might have a much smaller impact. Most recently I found that SqueezeNet takes 780ms to run in NCHW format and 182ms to run in NHWC format. It might be worth checking out your schedules on a squeezenet or vggnet (lots of standard convolutions).

Hi @jwfromm , thanks for you data. I have tried that schedule with MobileNetV1, but didn't get performance advantage like 5X. I think the performance gain depends on the network architecture also :)

**Comment by jackwish** [comment]
> Hey @jackwish , nice work for supporting NHWC data layout. It would be great if you could give some performance number of common models on ARM CPUs.

Thanks for the reminder @ZihengJiang , we will share some performance data when this is ready for review. :) We have been working on internal project, so the progress of this PR could be last for days :(

**Comment by jackwish** [comment]
> Hi @jackwish , any idea when this PR can be in good shape :)

Hi @anijain2305 , thank you for asking. This PR was pending due to NHWC building failure on ARM targets (w/o this patch). I didn't look into it in detail as I have been busy with Quantization and OpenCL backends. Will re-check later this week :)

**Comment by jackwish** [comment]
Rebased now. Collecting benchmarking data.

PS. Though we expected NHWC have a potential of better performance on CPU than NCHW, however, the result depends on how the schedule is designed. Say, a naive NHWC cannot outperform a carefully designed NCHW schedule.
We (AliOS team in Alibaba) have not observed better NHWC performance in spatial pack before, as not paying much effort in this. We was focusing on other schedules. So, this patch may not outperform the original NCHW spatial pack schedule. But still a good try :)

**Comment by anijain2305** [comment]
Thanks @jackwish for the update. Makes sense. So, are we planning to close this PR?

I am starting to look into Conv2D schedules for ARM. I can make a alter-layout change for NHWC network to use NCHW spatial schedule. I will do it in a manner that there are as few layout transforms as possible in the whole network (hopefully just 2) to amortize the cost of layout transforms.

**Comment by jackwish** [comment]
> Thanks @jackwish for the update. Makes sense. So, are we planning to close this PR?

I agree to close this - data layout doesn't impact spatial pack convolution very much. And people interesting in this can still try @cowanmeg 's [branch](https://github.com/cowanmeg/tvm/tree/quantized_end2end) or [mine](https://github.com/jackwish/tvm/tree/topi/conv-nhwc).

> I am starting to look into Conv2D schedules for ARM. I can make a alter-layout change for NHWC network to use NCHW spatial schedule. I will do it in a manner that there are as few layout transforms as possible in the whole network (hopefully just 2) to amortize the cost of layout transforms.

That sounds good! We have already considered this solution internally before.

Thanks everyone, this is now closing as we are not going to putting extra effort to improve the performance of Spatial Pack convolution schedule for NHWC.

**Comment by tqchen** [comment]
Given how far we have landed the code, it will still be great to bring it into the collection. @jackwish if you like, please feel free to re-open the thread and get some reviews.

**Comment by FrozenGene** [comment]
I agree @tqchen's comment. I think we could bring this in, which doesn't hurt anything.

**Comment by jackwish** [comment]
Per @tqchen and @FrozenGene suggestion, reopening.

As I didn't observe significant performance advantage of this NHWC spatial pack schedule, this is much like a functionality patch rather than something to outperform NCHW.

@tmoreau89 Would like to look into this code structure? I will add some benchmark results next week.

**Comment by snowolfhawk** [comment]
conv2d.py:625
why insert this new_weight below to overlay the previous new_weight? @jackwish
            new_weight = tvm.placeholder((KH + tile_size - 1, KH + tile_size -1, CO // VC, CI, VC),
                                         kernel.dtype)

**Comment by snowolfhawk** [comment]
conv2d.py:625
Another problem, the second KH should be KW!! @jackwish
new_weight = tvm.placeholder((KH + tile_size - 1, KH + tile_size -1, CO // VC, CI, VC),
kernel.dtype)

**Comment by jackwish** [comment]
Thank you for the review @snowolfhawk ! I will update later (I have been working on some other thing on parallel).

Btw, would you please to leave review comments by clicking the floating button below, that is pretty efficient :)
屏幕快照 2019-11-15 下午4 43 53

**Comment by snowolfhawk** [comment]
@jackwish
Do you still use winograd for NHWC?

**Comment by jackwish** [comment]
> Do you still use winograd for NHWC?

Hi @snowolfhawk , this Spatial Pack schedule seems to be the first NHWC layout support for ARM CPU. May I know more detail of your question?

**Comment by jackwish** [comment]
Rewritten schedule. For Mobilenet on Rasp 3B+ w/ AArch64 image, NCHW gets 134.26 ms while NHWC gets 132.50 ms. People may still finetune this schedule to improve.

Before this is merged, the NHWC->NCHW layout transform #4249 may need to revert, so please hold a second.

**Comment by FrozenGene** [comment]
> > Do you still use winograd for NHWC?
> 
> Hi @snowolfhawk , this Spatial Pack schedule seems to be the first NHWC layout support for ARM CPU. May I know more detail of your question?

I think @snowolfhawk 's question is we could use Winograd algo. on NCHW layout, For NCHW, could we use Winograd for NHWC layout too?

**Comment by anijain2305** [comment]
@jackwish I saw your comment regarding reverting https://github.com/apache/incubator-tvm/pull/4249

Sounds good to me. Lets ensure that we can work with both Conv and Depthwise. Let me know. If you want, I can send a PR for reverting.

**Comment by jackwish** [comment]
Hi @anijain2305 , actually that was an accident :(. I have mixed the data of different layouts... Will tune this when I have time...

**Comment by anijain2305** [comment]
Oh I see. No worries. I think AlterOpLayout code converted to NCHW under the hood, causing confusion.

It's ok. Take your time. I am willing to revert my PR if we see comparable performance with NHWC schedule. That will be the right thing to do.

**Comment by jackwish** [comment]
Rewritten schedule. For workload (input/output HW 64, input/output channel 64, kernel size 1), we now have comparable performance on Rasp 3B+ w/ AArch64 Ubuntu (~5ms). Tuning a larger workload currently.

So basically this is ready for review, please have a look at @tmoreau89 @ZihengJiang .

Notes:
* Kernel prepack not enabled yet. We'd like to hold that until depthwise NHWC were done, and the AlterOpLayout were simplified.
* Won't be used unless disable NHWC-NCHW layout transformation in AlterOpLayout. To try out this, put `return None` in AlterOpLayout.
* AlterOpLayout will be updated when we can run whole MobileNet in NHWC.

**Comment by jackwish** [comment]
For the [CI sanity check](http://ci.tvm.ai:8080/blue/organizations/jenkins/tvm/detail/PR-3859/5/pipeline) , I failed to reproduce it locally :(. And, I think the warning (as below) should be ignored because we are really using `cfg['compat'].val` as integer rather than bool.

```
python3 -m pylint topi/python/topi --rcfile=/workspace/tests/lint/pylintrc

Using config file /workspace/tests/lint/pylintrc

************* Module topi.arm_cpu.conv2d_spatial_pack

topi/python/topi/arm_cpu/conv2d_spatial_pack.py:314: [R1706(consider-using-ternary), schedule_conv2d_spatial_pack_nhwc] Consider using ternary (oco if cfg['compat'].val else owo)

topi/python/topi/arm_cpu/conv2d_spatial_pack.py:331: [R1706(consider-using-ternary), schedule_conv2d_spatial_pack_nhwc] Consider using ternary (oco if cfg['compat'].val else owo)



--------------------------------------------------------------------

Your code has been rated at 10.00/10 (previous run: 10.00/10, -0.00)
```

**Comment by FrozenGene** [comment]
@jackwish disable in current file is acceptable

File: topi/python/topi/arm_cpu/conv2d_spatial_pack.py [hunk 2]
```diff
@@ -196,3 +196,160 @@ def schedule_conv2d_spatial_pack_nchw(cfg, s, data_vec, kernel_vec,
...
+
+    # schedule conv
+    n, oho, owo, oco, ohi, owi, oci = s[conv].op.axis
+    ic, kh, kw = s[conv].op.reduce_axis
+    cfg['reorder_conv'].apply(s, conv, [n, oho, owo, oco, kh, kw, ohi, owi, ic, oci])
+    cfg['ann_reduce'].apply(s, conv, [kh, kw],
+                            axis_lens=[get_const_int(kh.dom.extent),
+                                       get_const_int(kw.dom.extent)],
+                            max_unroll=32,
+                            cfg=cfg)
```
Comment (FrozenGene) [inline topi/python/topi/arm_cpu/conv2d_spatial_pack.py]: max_unroll setting to be 32 maybe hang on some arm devices. The modest choices is 16 like NCHW schedule.

File: topi/python/topi/arm_cpu/conv2d_spatial_pack.py [hunk 3]
```diff
@@ -196,3 +196,160 @@ def schedule_conv2d_spatial_pack_nchw(cfg, s, data_vec, kernel_vec,
...
+                    axis=[ic, kh, kw]), name='conv')
+    else:
+        conv = tvm.compute(ovshape, lambda n, oho, owo, oco, ohi, owi, oci: \
+            tvm.sum(data_vec[n, oho, owo, ohi*HSTR+kh, owi*WSTR+kw, ic].astype(out_dtype) *
+                    kernel_vec[oco, kh, kw, ic, oci].astype(out_dtype),
+                    axis=[ic, kh, kw]), name='conv')
+
+    output = tvm.compute(oshape, lambda n, oho, owo, oc:
+                         conv[n][oho//OHI][owo//OWI][oc//OCI][oho%OHI][owo%OWI][oc%OCI],
+                         name='output_unpack', tag='spatial_conv_output_NHWC')
```
Comment (FrozenGene) [inline topi/python/topi/arm_cpu/conv2d_spatial_pack.py]: change operator // and % to tvm.indexdiv and tvm.indexmod. you could refer NCHW schedule

File: topi/python/topi/arm_cpu/conv2d_spatial_pack.py [hunk 4]
```diff
@@ -196,3 +196,160 @@ def schedule_conv2d_spatial_pack_nchw(cfg, s, data_vec, kernel_vec,
...
+        compat_axis = [owo, oco][cfg['compat'].val]
+        s[kernel_vec].compute_at(s[conv], compat_axis)
+        s[data_vec].compute_at(s[conv], compat_axis)
+
+    # schedule kernel pack
+    oco, kh, kw, ic, oci = kernel_vec.op.axis
+    s[kernel_vec].vectorize(oci)
+    s[kernel_vec].unroll(ic)
+    if cfg['compat'].val == 2:
+        s[kernel_vec].parallel(oco)
```
Comment (FrozenGene) [inline topi/python/topi/arm_cpu/conv2d_spatial_pack.py]: why we don’t use parallel directly?

File: topi/python/topi/arm_cpu/conv2d_spatial_pack.py [hunk 5]
```diff
@@ -196,3 +196,160 @@ def schedule_conv2d_spatial_pack_nchw(cfg, s, data_vec, kernel_vec,
...
+        s[data_vec].vectorize(owi)
+        s[data_vec].unroll(ohi)
+    else:
+        n, oho, owo, ohi, owi, ic = s[data_vec].op.axis
+        s[data_vec].vectorize(ic)
+        s[data_vec].unroll(owi)
+    if cfg['compat'].val == 2:
+        paxis = s[data_vec].fuse(n, oho)
+        s[data_vec].parallel(paxis)
+
```
Comment (FrozenGene) [inline topi/python/topi/arm_cpu/conv2d_spatial_pack.py]: same as kernel_vec question

File: topi/python/topi/arm_cpu/conv2d_spatial_pack.py (hunk 4 above)
Comment (jackwish) [reply topi/python/topi/arm_cpu/conv2d_spatial_pack.py]: Attaching `data_vec` and `kernel_vec` compute to `conv` when `cfg['compat'].val < 2`, in which case parallel them will lead to *nested parallel* error - `conv` has already been parallelized.

File: topi/python/topi/arm_cpu/conv2d_spatial_pack.py (hunk 5 above)
Comment (jackwish) [reply topi/python/topi/arm_cpu/conv2d_spatial_pack.py]: see https://github.com/apache/incubator-tvm/pull/3859#discussion_r349421208

**Comment by jackwish** [comment]
Thanks @FrozenGene for your quick review, comments addressed.

**Comment by tmoreau89** [comment]
@jackwish great work, thank you for pushing this PR through since it was opened. I'm glad the performance is on par on a small kernel. Do you mind checking tuned performance for e2e resnet18 and/or mobilenet? One thing to keep track of is the schedule space. If the schedule space for your implementation turns out to be very large, it might make tuning quite expensive. One thing to get around this is to use xgboost tuner, and limit the number of trials to something like 1000.

**Comment by tmoreau89** [comment]
Other than that, I'd recommend adding more comments in the schedule code. Particularly in terms of decisions that affect compute declaration/schedule space definition. For instance, why did you list the different dimension reorderings etc. This will help future maintainers, or folks who'd want to write future schedule templates for ARM.

**Comment by jackwish** [comment]
Hi @tmoreau89 , thanks for your comments!

> Do you mind checking tuned performance for e2e resnet18 and/or mobilenet?

I'd like to provide e2e performance too, though it seems pretty hard before we have a dedicated depthwise NHWC schedule. I will tune some workloads from MobileNet.

> One thing to keep track of is the schedule space.

The tuning space is a bit larger than NCHW, but still manageable. When evaluating the performance, I have a trail of 1000 and early stop of 500 for both NCHW and NHWC. The best result is obtained at about 500th iteration even with a sufficient large workload (image size 128 with input/output channel 128).

> I'd recommend adding more comments in the schedule code.

For sure I agree with this. Most of the TVM schedule primitives are straightforward (to people who have experience), while the recording are basically for locality in general regarding different workloads.

**Comment by tmoreau89** [comment]
@jackwish any updates on the e2e mobilenet tuning results? thanks!

**Comment by jackwish** [comment]
> @jackwish any updates on the e2e mobilenet tuning results? thanks!

@tmoreau89 I am very sorry for the delay, we have been working on enabling some models internally. I will update the data as soon as possible when available.

**Comment by tqchen** [comment]
What is the status of the PR, is it ready to be merged?

**Comment by FrozenGene** [comment]
@jackwish do you have spare time to continue to handle this? If we could support NHWC layout of depthwise convolution in this pr, it will be nice, because this will help users tune whole mobilenet when to use TFLite. I wish we could make it in ASAP.

**Comment by FrozenGene** [comment]
@jackwish If you don't have time to benchmark end2end results compared with NCHW, I think it doesn't matter, we could support this functionality and continue to evolve. @tqchen how about your opinion of this?

**Comment by tqchen** [comment]
Given the PR stands complete, i will merge it in for now, would be great if we can followup with a thread for benchmarks :) Thanks @jackwish @FrozenGene @tmoreau89 @snowolfhawk

**Comment by jackwish** [comment]
Thanks @tqchen for your help of merging this PR. The long progress have been interrupted many times due to busy internal work, I am sorry for that and related todo things. Currently, I lose access to the development environment as I have left the position, but I may still be available to help to improve related schedules. One this to for this PR is that, the *NHWC->NCHW layout transform* needs to be disabled to tune with this NHWC schedule. But please don't rush for it as I have not tested it with latest code.

Thank you everyone.
Best regards

**Pull request merged**